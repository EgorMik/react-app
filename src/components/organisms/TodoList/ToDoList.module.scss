$color: rgb(58, 59, 59);
$gradient: linear-gradient(to right, rgb(95, 95, 223), rgb(211, 125, 225));
$grayColor: #333;
$h: 10px;

@mixin horizontal-center() {
    margin: 0 auto;
}

@mixin clearfix {
    &:after {
        clear: both;
        content: '.';
        display: block;
        height: 0;
        line-height: 0;
        overflow: hidden;
    }
    *height: 1%;
}

@mixin border($borderWidth:1px,$borderColor:#d9d9d9,$borderStyle:solid) {
    border: $borderWidth $borderColor $borderStyle;
}

@mixin boxSizing {
    box-sizing: border-box;
    -moz-box-sizing: border-box;
    -webkit-box-sizing: border-box;
}

@mixin verticalCenter($height:$h*4) {
    height: $height;
    line-height: $height;
}

@mixin transform($transforms) {
    -moz-transform: $transforms;
    -o-transform: $transforms;
    -ms-transform: $transforms;
    -webkit-transform: $transforms;
    transform: $transforms;
}

@mixin rotate($deg) {
    @include transform(rotate(#{$deg}deg));
}



.list_box ul {
    list-style: none;
  }
  .list {
    margin-top:13px;
    width: 100%;
    background-color: white;
    border-radius: 0.8rem;
    box-shadow: 1px 1rem 3rem 11px rgba(0, 0, 0, 0.2);
}
.todo-list {
  display: block;
  background: rgb(251, 251, 250); 
  border-radius: 0.8rem;
  padding-left: 3px;
  @include clearfix;
  li {
      cursor: pointer;
      list-style: none;
      border-bottom: 1px solid #d9d9d9;
      width: 100%;
      @include clearfix;
      p {
          float: left;
          width: 84%;
          padding-left: 2%;
          @include verticalCenter(50px);
          margin: 0;
          cursor: pointer;
          color: grey;
      
      }
      
      span {
          width: 10%;
          @include verticalCenter(50px);
          text-align: right;
          padding-right: 10px;
          @include boxSizing;
          font-size: 18px;
          color: $color;
          cursor: pointer;
      }
  }
}