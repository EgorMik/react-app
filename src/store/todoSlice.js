import { createSlice} from '@reduxjs/toolkit';
export const filters = {
    ALL: "ALL",
    COMPLETED: "COMPLETED",
    NOT_COMPLETED: "NOT_COMPLETED"
  }
  const localStorageTodos = JSON.parse(localStorage.getItem('todo'));

if (!localStorageTodos) {
  const initalTodos = {
    todos: [ {
        id: '714ab407-82d7-4d96-ba8f-34c0f047acce',
        text: 'Complete online JavaScript course',
        completed: true,
      }],
  };

  localStorage.setItem('todo', JSON.stringify(initalTodos));
}

    
const todoSlice = createSlice({
    name:'todos',
    initialState:  JSON.parse(localStorage.getItem('todo')) !== null
    ? JSON.parse(localStorage.getItem('todo'))
    : { todos: [] },
    reducers:{
        addTodo(state, action){
            state.todos.push(action.payload)
            localStorage.setItem('todo', JSON.stringify(state));
       },
        removeTodo(state, action){
            state.todos = state.todos.filter(todo => todo.id !== action.payload.id);
           
            localStorage.setItem('todo', JSON.stringify(state));
        },
        toggleComplete(state, action){
            const toggledTodo = state.todos.find(todo => todo.id === action.payload.id);
            toggledTodo.completed = !toggledTodo.completed;
            localStorage.setItem('todo', JSON.stringify(state));
        },
        filterBy(state, action){
            state.filterBy = action.payload;
        },
        clearComplited(state, action){
            state.todos = state.todos.filter(todo => todo.completed !== action.payload);
            localStorage.setItem('todo', JSON.stringify(state));
        },
        dragAndDropSave(state, action) {
            state.todos = action.payload;
            localStorage.setItem('todo', JSON.stringify(state));
          },
    }
});
export const {addTodo,removeTodo,toggleComplete,filterBy,clearComplited,dragAndDropSave} = todoSlice.actions;
export default todoSlice.reducer;
