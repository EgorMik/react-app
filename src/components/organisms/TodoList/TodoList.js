import React from 'react';
import { useSelector, useDispatch} from 'react-redux';
import { filters, dragAndDropSave } from '../../../store/todoSlice';
import Todo from '../../molecules/Todo/Todo';
import Footer from '../../molecules/FooterList/Footer';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import styles from './ToDoList.module.scss';

const TodoList = () =>{
    let todos = useSelector((state) => state.todos.todos);
    const filter = useSelector((state) => state.todos.filterBy);
    const dispatch = useDispatch();
    
    const filteredTodo = () => {
      if(filter === filters.COMPLETED) {
        const newTodos = todos.filter(todo => todo.completed);
        return [...newTodos]
      }
        if(filter === filters.NOT_COMPLETED) {
        const newTodos = todos.filter(todo => !todo.completed)
        return [...newTodos]
      }
        return [...todos];
      };
    
      const onDragEndHandler = (result) => {
        const { destination, source } = result;
        if (!destination) {
          console.log(destination);
          return;
        }
        if (
          destination.droppableId === source.droppableId &&
          destination.index === source.index
        ) {
          return;
        }
        if (filter === "NOT_COMPLETED" || filter === "COMPLETED") {
          const todosUpdated = Array.from(filteredTodo());
          const [reorderedTodo] = todosUpdated.splice(source.index, 1);
          todosUpdated.splice(destination.index, 0, reorderedTodo);
          if (filter  === filters.NOT_COMPLETED) {
            const checked = todos.filter((todo) => todo.completed === true);
            const finallTodoList = checked.concat(todosUpdated);
            dispatch(dragAndDropSave(finallTodoList));
            return;
          }
        }
        const todosUpdated = Array.from(todos);
        const [reorderedTodo] = todosUpdated.splice(source.index, 1);
        todosUpdated.splice(destination.index, 0, reorderedTodo);
    
        dispatch(dragAndDropSave(todosUpdated));
      };
    return (
      <div className={styles.list}>
      <div className={styles.list_box}>
      <DragDropContext onDragEnd={onDragEndHandler}>
      <Droppable droppableId="todos">
      {(provided) => (
    <ul {...provided.droppableProps} ref={provided.innerRef} className={styles["todo-list"]}>
        {filteredTodo().map((value,index) => (
          <Todo
            key={value.id}
            id={value.id}
            completed={value.completed}
            text ={value.text}
            index={index}
          />
        ))}
    {provided.placeholder}
    </ul>
    )}
    </Droppable>
    </DragDropContext>
    </div>
      <Footer/>
    </div>    
);
};

export default TodoList;