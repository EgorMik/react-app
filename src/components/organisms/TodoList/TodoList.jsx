import React from 'react';
import { useSelector, useDispatch} from 'react-redux';
import { filters, dragAndDropSave } from '../../../store/todoSlice';
import Todo from '../../molecules/Todo/Todo';
import Footer from '../../molecules/FooterList/Footer';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import styles from './List.module.css';

const TodoList = ({displayStyle}) =>{
    const todos = useSelector(state => state.todos.todos);
    const filter = useSelector((state) => state.todos.filterBy);
    const dispatch = useDispatch();
    const filteredTodo = () => {
      if(filter === filters.COMPLETED) {
        const newTodos = todos.filter(todo => todo.completed);
        return [...newTodos]
      }
        if(filter === filters.NOT_COMPLETED) {
        const newTodos = todos.filter(todo => !todo.completed)
        return [...newTodos]
      }
        return [...todos];
      };
      const onDragEndHandler = (result) => {
        const { destination, source } = result;
        if (!destination) {
          console.log(destination);
          return;
        }
        if (
          destination.droppableId === source.droppableId &&
          destination.index === source.index
        ) {
          return;
        }
        if (filter === "NOT_COMPLETED" || filter === "COMPLETED") {
          const todosUpdated = Array.from(filteredTodo());
          const [reorderedTodo] = todosUpdated.splice(source.index, 1);
          todosUpdated.splice(destination.index, 0, reorderedTodo);
          if (filter  === filters.NOT_COMPLETED) {
            const checked = todos.filter((todo) => todo.completed === true);
            const finallTodoList = checked.concat(todosUpdated);
            dispatch(dragAndDropSave(finallTodoList));
            return;
          }
        }
        const todosUpdated = Array.from(todos);
        const [reorderedTodo] = todosUpdated.splice(source.index, 1);
        todosUpdated.splice(destination.index, 0, reorderedTodo);
    
        dispatch(dragAndDropSave(todosUpdated));
      };
    return (
      <div className="list">
      <div className={styles.list_box}>
      <DragDropContext onDragEnd={onDragEndHandler}>
      <Droppable droppableId="todos">
      {(provided) => (
    <ul {...provided.droppableProps} ref={provided.innerRef} className="dodo-list">
        {filteredTodo().map((item,index) => (
            
                <Todo
                    key={item.id}
                    id={item.id}
                    completed={item.completed}
                    text ={item.text}
                    index={index}
                />
        ))}
      {provided.placeholder}
    </ul>
    )}
    </Droppable>
    </DragDropContext>
    
      </div>
      <Footer displayStyle={displayStyle}/>
      </div>    
);
};
export default TodoList;