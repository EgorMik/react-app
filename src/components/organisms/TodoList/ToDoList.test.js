import React from "react";
import renderer from "react-test-renderer";
import TodoList from "./TodoList";
import { Provider } from "react-redux";
import store from '../../../store'
describe("TodoList", () => {
    it("should renders correctly", () => {
      const tree = renderer
        .create(
          <Provider store={store}>
            <TodoList />
          </Provider>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })});
