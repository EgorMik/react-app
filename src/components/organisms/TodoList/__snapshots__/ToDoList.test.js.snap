// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should renders correctly 1`] = `
<div
  className="list"
>
  <div
    className="list_box"
  >
    <ul
      className="todo-list"
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="todos"
    >
      <li
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="714ab407-82d7-4d96-ba8f-34c0f047acce"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="714ab407-82d7-4d96-ba8f-34c0f047acce"
        draggable={false}
        onDragStart={[Function]}
        onTransitionEnd={null}
        role="button"
        style={
          Object {
            "transform": null,
            "transition": null,
          }
        }
        tabIndex={0}
      >
        <input
          checked={true}
          className="checkbox"
          onChange={[Function]}
          onClick={[Function]}
          type="checkbox"
        />
        <p
          onClick={[Function]}
          style={
            Object {
              "color": "grey",
              "textDecoration": "line-through",
            }
          }
        >
          Complete online JavaScript course
        </p>
        <span
          className="deleteButton"
          onClick={[Function]}
        >
          x
        </span>
      </li>
    </ul>
  </div>
  <div
    className="footer"
  >
    <div>
      <p>
        0
         items left
      </p>
    </div>
    <div>
      <button
        className="unactive"
        onClick={[Function]}
      >
        All
      </button>
      <button
        className="unactive"
        onClick={[Function]}
      >
        Completed
      </button>
      <button
        className="unactive"
        onClick={[Function]}
      >
        Active
      </button>
    </div>
    <div>
      <button
        className="unactive"
        onClick={[Function]}
      >
        Clear Compleated
      </button>
    </div>
  </div>
</div>
`;
